cpp/src/replace/nulls.cu

```diff
@@ -352,6 +356,75 @@ std::unique_ptr<cudf::column> replace_nulls_scalar_kernel_forwarder::operator()<
   return cudf::dictionary::detail::replace_nulls(dict_input, replacement, stream, mr);
 }
 
+struct replace_nulls_replace_policy_functor {
+  __device__ thrust::tuple<cudf::size_type, bool> operator()(
+    thrust::tuple<cudf::size_type, bool> const& lhs,
+    thrust::tuple<cudf::size_type, bool> const& rhs)
+  {
+    return thrust::get<1>(rhs) ? thrust::make_tuple(thrust::get<0>(rhs), true)
+                               : thrust::make_tuple(thrust::get<0>(lhs), true);
+  }
+};
+
+/**
+ * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
+ *        `replace_nulls` with the appropriate data types.
+ */
+struct replace_nulls_replace_policy_kernel_forwarder {
+  template <typename col_type,
+            typename std::enable_if_t<cudf::is_fixed_width<col_type>()>* = nullptr>
+  std::unique_ptr<cudf::column> operator()(cudf::column_view const& input,
+                                           cudf::replace_policy const& replace_policy,
+                                           rmm::cuda_stream_view stream,
+                                           rmm::mr::device_memory_resource* mr)
+  {
+    using Type     = cudf::device_storage_type_t<col_type>;
+    auto device_in = cudf::column_device_view::create(input);
+    auto index     = thrust::make_counting_iterator<cudf::size_type>(0);
+    auto valid_it  = cudf::detail::make_validity_iterator(*device_in);
+    auto in_begin  = thrust::make_zip_iterator(thrust::make_tuple(index, valid_it));
+
+    rmm::device_vector<cudf::size_type> gather_map(input.size());
+    auto gm_begin = thrust::make_zip_iterator(
+      thrust::make_tuple(gather_map.begin(), thrust::make_discard_iterator()));
+
+    auto func = replace_nulls_replace_policy_functor();
+    if (replace_policy == cudf::replace_policy::PRECEDING) {
+      thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream.value()),
+                             in_begin,
+                             in_begin + input.size(),
+                             gm_begin,
+                             func);
+    } else {
+      auto in_rbegin = thrust::make_reverse_iterator(in_begin + input.size());
+      auto gm_rbegin = thrust::make_reverse_iterator(gm_begin + gather_map.size());
+      thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream.value()),
+                             in_rbegin,
+                             in_rbegin + input.size(),
+                             gm_rbegin,
+                             func);
+    }
+
+    cudf::column_view gm_col(cudf::data_type{cudf::type_id::INT32},
+                             gather_map.size(),
+                             thrust::raw_pointer_cast(gather_map.data()));
+
+    auto output =
+      cudf::gather(cudf::table_view({input}), gm_col, cudf::out_of_bounds_policy::DONT_CHECK, mr);
```

Comment:
Just use the `detail::gather` that works on iterators instead of constructing a `column_view` from the gather map.